File: src/problemSet/FindKthBitInNthBinaryString.java
package problemSet;

/**
 * 同{@link FindKthBitInNthBinaryStringKt}
 */
public class FindKthBitInNthBinaryString {
    public static void main(String[] args) {
        System.out.println(findKthBit(4, 11));
    }

    public static Character findKthBit(int n, int k) {
        return generateS(n).toCharArray()[k - 1];
    }

    public static String generateS(int n) {
        if (n == 1) {
            return "0";
        }
        String v = generateS(n - 1);
        return v + "1" + reverse(invert(v));
    }

    public static String reverse(String str) {
        return new StringBuilder(str).reverse().toString();
    }

    public static String invert(String str) {
        StringBuilder stringBuilder = new StringBuilder(str);
        for (int i = 0; i < stringBuilder.length(); i++) {
            stringBuilder.setCharAt(i, stringBuilder.charAt(i) == '1' ? '0' : '1');
        }
        return stringBuilder.toString();
    }
}

File: src/problemSet/FindKthBitInNthBinaryString.kt
package problemSet

/**
 * 周赛201
 * 1545. 找出第 N 个二进制字符串中的第 K 位
 * medium
 *
 * https://leetcode-cn.com/contest/weekly-contest-201/
 * https://leetcode-cn.com/problems/find-kth-bit-in-nth-binary-string/
 *
 * Kotlin
 * https://leetcode-cn.com/submissions/detail/124650507/
 * 由于貌似没人用kotlin写这道题，所以又用Java写了一遍[FindKthBitInNthBinaryString]
 * Java
 * https://leetcode-cn.com/submissions/detail/124653203/
 */
fun main() {
    println(findKthBit(4, 11))
}

fun findKthBit(n: Int, k: Int): Char {
    return generateS(n)[k - 1]
}

fun generateS(n: Int): String {
    if (n == 1) {
        return "0"
    }
    val v = generateS(n - 1)
    return v.plus("1").plus(reverse(invert(v)))
}

fun reverse(str: String): String {
    return str.reversed()
}

fun invert(str: String): String {
    val stringBuilder = StringBuilder(str)
    stringBuilder.forEachIndexed { index, c ->
        stringBuilder[index] = if (c == '1') '0' else '1'
    }
    return stringBuilder.toString()
}